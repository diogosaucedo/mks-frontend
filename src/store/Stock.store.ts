import { PayloadAction, createSlice } from '@reduxjs/toolkit';

import { ICardProduct } from '@/interfaces';

const stock = createSlice({
  name: 'stock',
  initialState: {
    counter: 0,
    total: 0,
    visible: false,
    products: [] as never | ICardProduct[],
  },
  reducers: {
    toggleVisible(state) {
      state.visible = !state.visible;
    },
    addProduct(state, action: PayloadAction<ICardProduct>) {
      if (
        state.products.find(
          (item: ICardProduct) => item.id == action.payload.id,
        )
      ) {
      } else {
        state.products.push(action.payload);
      }
    },
    removeProduct(state, action: PayloadAction<number>) {
      state.products = state.products.filter(
        (product) => product.id !== action.payload,
      );
    },
    counter(state) {
      state.counter = state.products.length;
    },
    getTotal(state) {
      state.total = state.products.reduce(
        (acc, item) => acc + item.amount * item.price,
        0,
      );
    },
    incrementAmountProduct(state, action: PayloadAction<number>) {
      state.products = state.products.map((product) => {
        if (product.id === action.payload) {
          return { ...product, amount: product.amount + 1 };
        } else {
          return product;
        }
      });
    },
    decrementAmountProduct(state, action: PayloadAction<number>) {
      state.products = state.products.map((product) => {
        if (product.id === action.payload && product.amount > 1) {
          return { ...product, amount: product.amount - 1 };
        } else {
          return product;
        }
      });
    },
  },
});

export const {
  toggleVisible,
  addProduct,
  counter,
  removeProduct,
  getTotal,
  incrementAmountProduct,
  decrementAmountProduct,
} = stock.actions;
export default stock.reducer;
