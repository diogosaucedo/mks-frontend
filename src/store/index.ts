import { configureStore } from '@reduxjs/toolkit'
import stockStore from './Stock.store'

export const store = configureStore({
  reducer: {
    stock: stockStore
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
