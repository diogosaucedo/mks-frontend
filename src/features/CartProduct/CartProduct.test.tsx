import { store } from '@/store/';
import {
  removeProduct,
  counter,
  incrementAmountProduct,
  decrementAmountProduct,
  getTotal,
} from '@/store/Stock.store';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider, useDispatch } from 'react-redux';

import { CartProduct } from '.';

jest.mock('react-redux', () => {
  const originalModule = jest.requireActual('react-redux');

  return {
    ...originalModule,
    useDispatch: jest.fn().mockReturnValue(jest.fn()),
  };
});

describe('CardProduct feature', () => {
  it('should render properties correctly', () => {
    const product = {
      id: 8,
      name: 'Headset Cloud Stinger',
      photo:
        'https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp',
      price: 600.0,
      amount: 1,
    };

    const truncatedPrice = Math.trunc(product.price);

    render(
      <Provider store={store}>
        <CartProduct key={product.id} data={product} />
      </Provider>,
    );
    expect(screen.getByText(product.name)).toBeInTheDocument();
    expect(screen.getByText(`R$${truncatedPrice}`)).toBeInTheDocument();
    expect(screen.getByAltText(product.name)).toBeInTheDocument();
    expect(screen.getByAltText(product.name)).toHaveAttribute(
      'src',
      product.photo,
    );
  });

  it('should match snapshot', () => {
    const product = {
      id: 8,
      name: 'Headset Cloud Stinger',
      photo:
        'https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp',
      price: 600.0,
      amount: 1,
    };

    const { asFragment } = render(
      <Provider store={store}>
        <CartProduct key={product.id} data={product} />
      </Provider>,
    );

    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('should dispatch actions correctly', () => {
    const dispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(dispatch);

    const product = {
      id: 8,
      name: 'Headset Cloud Stinger',
      photo:
        'https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp',
      price: 600.0,
      amount: 1,
    };

    render(
      <Provider store={store}>
        <CartProduct key={product.id} data={product} />
      </Provider>,
    );

    const buttonIncrement = screen.getByText('+');
    buttonIncrement.click();

    expect(dispatch).toHaveBeenCalledTimes(3);
    expect(dispatch).toHaveBeenCalledWith(incrementAmountProduct(product.id));
    expect(dispatch).toHaveBeenCalledWith(counter());
    expect(dispatch).toHaveBeenCalledWith(getTotal());

    const buttonDecrement = screen.getByText('-');
    buttonDecrement.click();

    expect(dispatch).toHaveBeenCalledTimes(6);
    expect(dispatch).toHaveBeenCalledWith(decrementAmountProduct(product.id));
    expect(dispatch).toHaveBeenCalledWith(counter());
    expect(dispatch).toHaveBeenCalledWith(getTotal());

    const buttonRemove = screen.getByText('X');
    buttonRemove.click();

    expect(dispatch).toHaveBeenCalledTimes(9);
    expect(dispatch).toHaveBeenCalledWith(removeProduct(product.id));
    expect(dispatch).toHaveBeenCalledWith(counter());
    expect(dispatch).toHaveBeenCalledWith(getTotal());
  });
});
