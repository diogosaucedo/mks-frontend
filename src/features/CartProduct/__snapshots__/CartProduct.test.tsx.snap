// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardProduct feature should match snapshot 1`] = `
<div
  class="sc-kDvujY kPajjf sc-ipEyDJ cIDRcz"
>
  <div
    class="sc-jSUZER klzDyK"
  >
    <div
      class="sc-gKPRtg fNtGPD"
    >
      <img
        alt="Headset Cloud Stinger"
        class="sc-ftTHYK jzygvA"
        height="60px"
        src="https://mks-sistemas.nyc3.digitaloceanspaces.com/products/hyperxcloudstinger.webp"
        width="fit-content"
      />
      <p
        class="sc-gswNZR ianvmX"
        font-size="13px"
        width="100px"
      >
        Headset Cloud Stinger
      </p>
    </div>
    <div
      class="sc-jSUZER bIGWDn"
    >
      <div
        class="sc-jSUZER kXwoZG"
      >
        <p
          class="sc-gswNZR fcuUFZ"
          font-size="5px"
        >
          Qtd:
        </p>
        <div
          class="sc-jSUZER fwiXev"
        >
          <button
            class="sc-hLBbgP cIfbAO"
            font-size="12px"
            height="19px"
            width="15px"
          >
            -
          </button>
          <input
            class="sc-eDvSVe iAslpS"
            disabled=""
            font-size="12px"
            height="19px"
            type="text"
            value="1"
            width="20px"
          />
          <button
            class="sc-hLBbgP cIfbAO"
            font-size="12px"
            height="19px"
            width="15px"
          >
            +
          </button>
        </div>
      </div>
      <p
        class="sc-gswNZR dRomFG sc-dkrFOg kJtyTs"
        font-size="14px"
        font-weight="700"
      >
        R$600
      </p>
    </div>
  </div>
  <button
    class="sc-hLBbgP gdtDKI"
    height="18px"
    transform="translate(10px, -45px)"
    width="18px"
  >
    <p
      class="sc-gswNZR bOUCb"
      color="#FFFFFF"
      font-size="8px"
    >
      X
    </p>
  </button>
</div>
`;
